### ISSUE TITLE
behaviour of offerToReceive* set to false when there is a local track

### ISSUE THREAD

**Issue by fippo** [w3c/webrtc-pc#1586]
related to #1383 but that is complicated enough already.
Suppose I do this:
```
var pc = new RTCPeerConnection();
navigator.mediaDevices.getUserMedia({audio: true})
.then(stream => {
  pc.addTrack(stream.getAudioTracks()[0], stream);
  return pc.createOffer({offerToReceiveAudio: false})
})
.then(offer => {
  console.log(offer.sdp);
});
```
I want to send audio but not receive it. This means the sdp should say "sendonly".
But we currently specify this as "leave it alone".

This needs something like "if there is a transceiver for this kind set its direction to sendonly".

**Comment by fippo** [comment]
I knew this was somewhat familiar, see #1461 and #1466. But both were closed leaving the poor guy trying to shim this 😭

**Comment by stefhak** [comment]
I feel your pain, but I wonder, was ```{offerToReceiveAudio: false}``` ever defined to make the m-line say ```sendonly```? My recollection says the use of ```offerToReceiveX``` was only to make the offer contain an m-line of kind ```X``` even if there was no outgoing media of type ```X```.

**Comment by jan-ivar** [comment]
> was {offerToReceiveAudio: false} ever defined to make the m-line say sendonly

@stefhak The spec [used to say](https://dev.w3.org/2011/webrtc/editor/archives/20130603/webrtc.html#constraints): *"This constraint allows an application to indicate its preferences for receiving video when creating an offer."*

and 

*"This is an enum type constraint that can take the values "true" and "false". The default is a non mandatory "true" for an RTCPeerConnection object that has a video stream at the point in time when the constraints are being evaluated and is non mandatory "false" otherwise."*

I interpreted it at the time to mean that, which is why Firefox works that way. Chrome does too, as [I pointed out](https://github.com/w3c/webrtc-pc/issues/1461#issue-241153792).

**Comment by jan-ivar** [comment]
What convinced me to not pursue this is @taylor-b's [comment](https://github.com/w3c/webrtc-pc/pull/1466#issuecomment-315227981). The scope of createOffer isn't to discover whom you're dealing with, rather just to discover which machines will be involved. Specifically: apps are generally on both sides of a connection, and should know which side intends to send what ahead of OA.

I could be wrong about this, of course. So others feel free to raise this issue if there's an important use-case left in the cold by this.

**Comment by taylor-b** [comment]
I'd suggest leaving this alone. The main reason for adding `offerToReceive` to the legacy section was for backwards compatibility, so that the common case of using it to trigger the generation of an "m=" section continues to work. If an application needs to change the directional attribute, that's more uncommon, and it can use `setDirection`.

**Comment by fippo** [comment]
Looking back at the number of "i want to broadcast" requests for simplewebrtc I think offerToReceive=false is pretty common.

> Specifically: apps are generally on both sides of a connection, and should know which side intends to send what ahead of OA.

You assume single-purpose apps. More complex applications that support both audiovideo and audio-only will probably need to signal the intent alongside the SDP. In which case its easier to just do it in the SDP.

3/3 browsers that implement legacy constraints set sendonly in the sdp (and I have a shim which should do it for the one that does not implement them) when there is a local track. They are probably not going to change that just because we avoid specifying it. Also beware of [the dragons](https://github.com/rtcweb-wg/jsep/pull/833)

Either way, I can fall back to "do what the others do" when there is no specification ;-)

**Comment by jan-ivar** [comment]
> In which case its easier to just do it in the SDP.

A sentence I never thought I'd hear! :)

**Comment by stefhak** [comment]
I realize that JSEP for a long time said that offerToReceive = 0/false was indeed a way to generate sendonly m-lines (my memory was for an even older version of JSEP (dated July 2014!)) - so I was wrong in https://github.com/w3c/webrtc-pc/issues/1586#issuecomment-328338926.

**Comment by aboba** [comment]
RESOLUTION from TPAC: we don't change how offerToReceive*: false behaves in the spec
https://www.w3.org/2017/11/06-webrtc-minutes.html

**Comment by fippo** [comment]
I do not see anything in the minutes that explains the decision. Is the WG intention to ship an utterly broken  "legacy" support?
Given that the transceivers direction needs to be specified anyway (https://github.com/w3c/webrtc-pc/issues/1662#issuecomment-342727739) creating a transceiver and then setting its direction does not seem like much extra effort.

The offer to make a PR still stands.

**Comment by stefhak** [comment]
@fippo of course you can submit a PR, and we will see what the group thinks of it. The resolution at TPAC was however to not do this.
Looking at the commits in fippo/webrtc-pc I wonder: is there such a thing as a _first_ track or transceiver?

**Comment by fippo** [comment]
@stefhak yes, that point still needs some work.
[jsep](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-20#section-5.2) has the following text which implies there is a 'first' even if the algorithm to collect transceivers does not have any ordering:
`This is done in the order the RtpTransceivers were added to the PeerConnection`

**Comment by stefhak** [comment]
@fippo I note that not only would we need order (to know which is 'first'), but also keep track of which transceivers are 'new' as the behavior specified by [JSEP a long time ago](https://github.com/w3c/webrtc-pc/issues/1586#issuecomment-3460708949) says the setting does not apply to negotiated transceivers.

JSEP at the time however said that all 'fresh' transceivers with an associated outgoing track should be marked as ```sendonly``` if offerToRecieveX is either false or 0, so that seems to take away the need to identify which is 'first'.

How many implementations support this today?

**Comment by fippo** [comment]
@stefhak uuugh -- broken link though.

> How many implementations support this today?

https://jsfiddle.net/mc5j66zk/2/ passes in Chrome, Firefox, Edge (with rtcpeerconnection-shim) and Edge's native "1.0" (if you fix the constructor issues and use addStream). Safari does neither implement this nor is it [shim-able currently](https://github.com/webrtc/adapter/pull/676)

I am somewhat keen to see how many of these implementations allow a subsequent offer with offerToReceive set to false ignoring what you/jsep requires about not applying to negotiated transceivers ;-)

**Comment by fippo** [comment]
so I found out that offerToReceiveAudio: false acts on *all* mlines in Firefox stable (57, pre-transceiver). This makes sense, it creates an offer that does *not* want to receive audio.

On the other hand offerToReceiveAudio: true only applies to the first non-stopped transceiver. This is a weird API.

**Comment by jan-ivar** [comment]
@fippo It makes more sense when you consider the boolean used to be a number. 0 gives you zero received media in spite of what you're sending, and 1 gives you one sent media in spite of what you're receiving.

**Comment by fippo** [comment]
yeah. The boolean only made sense with a single m-line (per-kind). Unified plan was probably the reason to go for integers. But now we have mixed up booleans with that and... it is a mess ;-)

**Comment by jan-ivar** [comment]
It's a legacy appendage. Doesn't have to be pretty. You're supposed to ignore it.

**Comment by fippo** [comment]
fixed in #1683
